### PR TITLE
k-d tree (A k-dimensional tree)

### PR DESCRIPTION
**Pull request by bomenderick**
## Description
A K-D Tree(also known as a K-Dimensional Tree) is a binary search tree where data in each node is a K-Dimensional point in space. In short, it is a space partitioning data structure for organizing points in a K-Dimensional space in other to facilitate nearest neighbor search of points.
In addition to insert, search, and delete methods, the implementation also supports nearest neighbors search, median finding for insertion in other to keep the k-d tree balanced, and a merge method to combine two k-d trees by collecting their points and building a balanced k-d tree from it.

Read more: 
- https://www.geeksforgeeks.org/search-and-insertion-in-k-dimensional-tree/
- https://www.geeksforgeeks.org/deletion-in-k-dimensional-tree/

### CODE REVIEW

File: src/data_structures/kd_tree.rs [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+
+impl<T: PartialOrd + Copy, const K: usize> Default for KDTree<T, K> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: PartialOrd + Copy, const K: usize> KDTree<T, K> {
+    // Create and empty kd-tree
+    // #[must_use]
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:39]:
```suggestion
```

File: src/data_structures/kd_tree.rs [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+        KDTree {
+            root: None,
+            size: 0,
+        }
+    }
+
+    // Returns true if point found, false otherwise
+    pub fn contains(&self, point: &[T; K]) -> bool {
+        search_rec(&self.root, point, 0)
+    }
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:50]: Is this really needed?

File: src/data_structures/kd_tree.rs [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+    // Returns true if successfully delete a point, false otherwise
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:52]: delete?

File: src/data_structures/kd_tree.rs [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+    where
+        T: Sum + Signed + Real,
+    {
+        let mut neighbors: Vec<(T, [T; K])> = Vec::new();
+        n_nearest_neighbors(&self.root, point, n, 0, &mut neighbors);
+        neighbors
+    }
+
+    // Returns the number of points in a kd-tree
+    // #[must_use]
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:81]:
```suggestion
```

File: src/data_structures/kd_tree.rs [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+    pub fn len(&self) -> usize {
+        self.size
+    }
+
+    // Returns the depth a kd-tree
+    // #[must_use]
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:87]:
```suggestion
```

File: src/data_structures/kd_tree.rs [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+    pub fn depth(&self) -> usize {
+        depth_rec(&self.root, 0, 0)
+    }
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:90]: Is this needed?

File: src/data_structures/kd_tree.rs [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+    // Determine whether there exist points in a kd-tree or not
+    // #[must_use]
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:93]:
```suggestion
```

File: src/data_structures/kd_tree.rs [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+    pub fn is_empty(&self) -> bool {
+        self.root.is_none()
+    }
+
+    // Returns a kd-tree built from a vector points
+    // #[must_use]
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:99]:
```suggestion
```

File: src/data_structures/kd_tree.rs [hunk 9]
```diff
@@ -0,0 +1,470 @@
...
+            tree.size = points.len();
+            tree.root = build_rec(points, 0);
+
+            tree
+        }
+    }
+
+    /// Returns a `KDTree` containing both trees
+    /// Merging two KDTrees by collecting points and rebuilding
+    // #[must_use]
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:114]:
```suggestion
```

File: src/data_structures/kd_tree.rs [hunk 10]
```diff
@@ -0,0 +1,470 @@
...
+    pub fn merge(&mut self, other: &mut Self) -> Self {
+        let mut points: Vec<[T; K]> = Vec::new();
+        collect_points(&self.root, &mut points);
+        collect_points(&other.root, &mut points);
+        KDTree::build(points)
+    }
```
Comment (vil02) [inline src/data_structures/kd_tree.rs:120]: Is this really needed?

**Review by vil02 — CHANGES_REQUESTED** [review]
@bomenderick: thanks for interesting contribution.

Before this will be merged some work needs to be done. First of all: please add missing tests. Especially the functionality finding the closest point has to be exercised quite brutally.

Could you also describe/document in the code, which operations does your implementation support?

File: src/data_structures/kd_tree.rs (hunk 2 above)
Comment (bomenderick) [reply src/data_structures/kd_tree.rs:50]: `Contains` is useful to search for the presence of a point in a k-d tree.
Or do you mean the additional call of `search_rec`?

Comment (vil02) [reply src/data_structures/kd_tree.rs:71]: Does `contains` need to be public?

Comment (bomenderick) [reply src/data_structures/kd_tree.rs:71]: **Yes** for interaction with the outside world. However, `contains` is just to complete `CRUD` operations on a k-d tree.
Indeed, the current k-d tree implementation doesn't make use of `contains`. However, its implementation could help as it is just a synonym of the `search` method of a k-d tree.
Would you recommend I make it private for now or remove it?

**Pull request closed without merging**